### PR TITLE
Subworkflow: parallelise vep

### PR DESCRIPTION
**Pull request by matthdsm**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by matthdsm** [comment]
@maxulysse , @nvnieuwk, time to start tearing this down!

**Comment by nvnieuwk** [comment]
gif

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
Good luck 😁

File: subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf [hunk 1]
```diff
@@ -2,12 +2,14 @@
...
 include { ENSEMBLVEP_VEP } from '../../../modules/nf-core/ensemblvep/vep/main'
 include { TABIX_TABIX    } from '../../../modules/nf-core/tabix/tabix/main'
 
 workflow VCF_ANNOTATE_ENSEMBLVEP {
     take:
-    vcf               // channel: [ val(meta), vcf ]
+    vcf               // channel: [ val(meta), vcf, tbi ]
     fasta             //   value: fasta to use (optionnal)
     vep_genome        //   value: genome to use
     vep_species       //   value: species to use
```
[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:15]: so fasta is no longer optionnal?
And do we really need tbis?

File: subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf [hunk 2]
```diff
@@ -18,13 +20,54 @@ workflow VCF_ANNOTATE_ENSEMBLVEP {
...
-    ENSEMBLVEP_VEP(vcf, vep_genome, vep_species, vep_cache_version, vep_cache, fasta, vep_extra_files)
-    TABIX_TABIX(ENSEMBLVEP_VEP.out.vcf)
+    BCFTOOLS_SPLIT(vcf)
+    ch_split_vcf = BCFTOOLS_SPLIT.out.split_vcf
+        .map{meta, vcf ->
+            meta.chunks = vcf instanceof List ? vcf.size() : 1
+            vcf_list = vcf instanceof List ? vcf : [vcf]
+            return [meta, vcf_list]
+        }
+        .transpose()
```
[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:30]: why creating a `new_meta` if you're not using it

File: tests/subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf [hunk 3]
```diff
@@ -5,10 +5,11 @@ nextflow.enable.dsl = 2
 include { VCF_ANNOTATE_ENSEMBLVEP } from '../../../../subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf'
 
 workflow vcf_annotation_ensemblvep {
-    input = [
+    input = Channel.of([
```
Comment (maxulysse) [inline tests/subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:8]: <3

File: subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf (hunk 1 above)
[the remark this answers is not in the document]
Comment (nvnieuwk) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:15]: Fixed!

File: subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf (hunk 2 above)
[the remark this answers is not in the document]
Comment (nvnieuwk) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:30]: I must have been distracted while reviewing this, sorry 😁, I'll change it!

[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:30]: why calling it `new_meta` when it's actually only the number of elements in the list?

[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:30]: I'd rather we do:
```suggestion
    BCFTOOLS_SPLIT(ch_vcf)
    ch_split_vcf = BCFTOOLS_SPLIT.out.split_vcf
        .map{ meta, vcf -> [ meta + [ size:vcf instanceof List ? vcf.size() : 1 ], vcf instanceof List ? vcf : [ vcf ] ] }.transpose()
```

[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:30]: whit of course some added comments

**Comment by maxulysse** [comment]
> The code itself looks fine, but you should add some explanation about the chunking in the docs. VEP has a large overhead to running which will affect the number of chunks (i.e., anyone using cloud processing will want to not split).

+1 can we had a boolean as an input to this subworkflow to disable splitting?

File: subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf [hunk 4]
```diff
@@ -18,13 +20,49 @@ workflow VCF_ANNOTATE_ENSEMBLVEP {
...
+        vep_genome,
+        vep_species,
+        vep_cache_version,
+        vep_cache,
+        fasta,
+        vep_extra_files
+    )
+    ch_split_annotated_vcf = ENSEMBLVEP_VEP.out.vcf
+        .map{meta, vcf ->
+            return [groupKey(meta - meta.subMap('chunks'), meta.chunks), vcf]
```
Comment (maxulysse) [inline subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf:38]:
```suggestion
            return [groupKey(meta - meta.subMap('size'), meta.size), vcf]
```

**Comment by matthdsm** [comment]
Actually, I'm considering dropping this altogether, since @nvnieuwk had a much better idea wrt parallelisation.
We can scan the tbi for contigs and then give those to vep to limit the regions. That way we avoid a bunch of IO and file shuffling

👍🏻  if you're pro
👎🏻  if you're against

File: subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf]: `count`?

[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf]: `split_count`

[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf]: `vcf_split_size`?

[the remark this answers is not in the document]
Comment (nvnieuwk) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf]: I like that one!

[the remark this answers is not in the document]
Comment (maxulysse) [reply subworkflows/nf-core/vcf_annotate_ensemblvep/main.nf]: as long we remove it after the subworkflows, I'm happy

**Comment by matthdsm** [comment]
> You're assuming quite a lot by splitting per contig
Well we're kind of doing the same right now with splitting the file by chr, no?

**Comment by matthdsm** [comment]
For those interested: https://github.com/bioconda/bioconda-recipes/pull/39868

**Pull request closed without merging**